### PR TITLE
add htmldom.py + a bit of the html tutorial

### PR DESCRIPTION
**Pull request by rmorshea**
This includes an intro for html tags (up until I got distracted making `htmldom.py`).

`htmldom.py` implements a more complete version of the `html` tag we discussed in person - it supports attribute expansion along with interpolation of tags and attributes and does so while still relying on `html.parser`. For example, the following:

```python
color = "blue"
attrs = {"style": {"font-size": "bold", "font-family": "mono"}}
dom = html"<a {attrs} color=dark{color} >{html'<h{i}/>' for i in range(1, 4)}</a>".render(indent=2)
print(dom)
```

Produces:

```
<a style="font-size:bold; font-family:mono" color="darkblue">
  <h1></h1>
  <h2></h2>
  <h3></h3>
</a>
```

My hope was that if I could implement this more fully featured `html` tag using `html.parser` it would remain approachable while also being more compelling. However, having implemented it now, I'm unsure of how it will pan out given the added complexity.

### CODE REVIEW

**Comment by rmorshea** [comment]
`htmldom.py` includes some editorializing. Specifically, it disallows attribute name interpolation. My reasoning here is that if you truly need to interpolate the name of an HTML attribute, you could just as easily form the name in a dictionary, and then use attribute expansion:

```python
attrs = {f"my{attr}": True}
html"<a {attrs} />"
```

At present, I'm unsure if JSX has come to the same decision.

File: examples/htmldom.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                case True:
+                    attribute_list.append(f" {key}")
+                case False | None:
+                    pass
+                case value:
+                    attribute_list.append(f' {key}="{escape(str(value))}"')
+
+        children_list: list[str] = []
+        for item in self.children:
+            if isinstance(item, str):
```
Comment (jimbaker) [inline examples/htmldom.py:62]: Use structural pattern matching here, instead of if/elif/else

**Review by jimbaker — COMMENTED** [review]
Looks like a good start!

I think your approach is quite good, because it explicitly marks the interpolation with `${...}` during feeding data, then processes the interpolation properly given the context in the handlers. So this doesn't rely on any `HTMLParser` internals, but it also enables caching the template.

I think it's possible to simplify this presentation in a nice literate fashion by showing the high-level aspects first, then filling in other parts as we go.

**Comment by rmorshea** [comment]
Thanks for the feedback. I'll make those changes and see if it's possible to break down the program in a reasonable way for the tutorial.

**Comment by rmorshea** [comment]
I've managed to simplify `htmldom.py` even more. I realized that I could store all the expression values in a list and interleave them with the HTML as it was parsed. This removes the need for any hackery with `string.Template` and will make things much easier to explain in the tutorial (which I started getting to the meat of).

File: examples/htmldom.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    interleaved_values: list[str] = []
+    for s, v in zip(string_parts[:-1], values):
+        interleaved_values.append(s)
+        interleaved_values.append(v)
+    interleaved_values.append(string_parts[-1])
+
+    return interleaved_values, remaining_values
+
+
+def _format_value(value: Any, conv: str, spec: str) -> Any | str:
```
Comment (jimbaker) [inline examples/htmldom.py:174]: See https://github.com/jimbaker/tagstr/blob/main/examples/taglib.py#L27 - this functionality is going to be generic, so we should have it in taglib

**Review by jimbaker — APPROVED** [review]
Looks good! (Sorry, I was out of band due to holidays.)

I really like seeing the id generation, this is a key aspect of what make this templating approach not just interesting/cool/Pythonic, but addressing a key problem in working with complex, modern templating requirements.

One idea I had in reading this is we may may want to emphasize *earlier* the last point about how tags can make HTML templates more readable, because we can structure them with ordinary Python coding practices. Maybe the best thing to do is to announce this goal at the beginning of the section - here's what you will be able to do, it's going to look like this - then work to it through the same set of steps.

Given this is still very much a draft overall for the tutorial, I will approve this now so we can work against it in main - much easier to improve flow there.

**Comment by rmorshea** [comment]
Definitely agree. I think nearly all the content is there, but the into "Applications in templating" section does not cleanly lead into the meat of the tutorial in the following "Writing an html tag" section. I'll work on follow-up edits.

**Pull request merged**